### PR TITLE
Add a pre-process method

### LINKED ISSUE #25: Add a pre-process step
## Description

In order to allow for inclusion of stationlist type settings into the yaml file I have a suggestion that we include a pre-process step.  This step would be run at the start of the yaml processor and would just cat the contents of a file at a suitable hierarchy.

A suggested use would be to split a file such as:
https://github.com/MetOffice/jjdocs/blob/develop/docs/share/jopa/um-gl/gl_ops-um-jedi_process_iasi.yaml

Into one file which includes lines 1-117.  This would go into jjaux and be called something like iasi_anchors.yaml.  The current file would then have something like:

``` yaml
PREPROCESS: <file path to file in jjaux>
```

This would then explicitly separate out the types of includes from a parsed file to a straight include.

The alternative I was thinking of would be to cat the files in the shell script which runs the task.  This would be less explicit than using the yaml parser.

### PR DESCRIPTION
**Pull request by mikecooke77**
## Description

This PR addresses the second part of #25 by adding a new method which can be called using `yp-preprocessor`.

## Issue(s) addressed

Resolves #25

## Dependencies

No dependencies.

## Impact

None

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have run the unit tests before creating the PR

### CODE REVIEW

File: src/yamlprocessor/datapreprocessor.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Process input and output "
+                    "files with multiple --define options."
+    )
+
+    # Positional argument for input
+    parser.add_argument('input_file', type=str, help='Input file')
```
Comment (DavidSimonin) [inline src/yamlprocessor/datapreprocessor.py:78]: Should this be:
`parser.add_argument('--input_file', '-i', type=str, help='Input file')`

File: src/yamlprocessor/datapreprocessor.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        action="store",
+        help='Name of output file, "-" for STDOUT'
+    )
+
+    # Optional --define arguments
+    parser.add_argument(
+        '--define',
+        action='append',
+        help='Key-value pairs in the format key=value', default=[]
+    )
```
Comment (DavidSimonin) [inline src/yamlprocessor/datapreprocessor.py]: Should this be:

```
parser.add_argument(
        '--define', '-d'
        action='append',
        help='Key-value pairs in the format key=value', default=[]
    )

```

File: docs/data-preprocessor.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+.. code-block:: yaml
+
+   DIRECT_INCLUDE=/path/to/file/to/be/included
+
+Command line useage
+-------------------
+
+.. code-block:: bash
+
+   yp-preprocessor [options] -o output-file-name input-file-name
```
Comment (DavidSimonin) [inline docs/data-preprocessor.rst]: Shouls this be:
` yp-preprocessor -i  input-file-name -o output-file-name [options]`

File: docs/cli.rst [hunk 4]
```diff
@@ -69,6 +69,30 @@ See :doc:`data-process` for detail.
...
    See also :envvar:`YP_TIME_REF_VALUE`.
 
+yp-preprocesor
+--------------
+
+Usage:
+
+.. code-block:: bash
+
+   yp-preprocesor [options] -o output-file-name input-file-name
```
Comment (DavidSimonin) [inline docs/cli.rst]: Shouls this be:
` yp-preprocessor -i  input-file-name -o output-file-name [options]`

File: docs/cli.rst [hunk 5]
```diff
@@ -69,6 +69,30 @@ See :doc:`data-process` for detail.
...
+
+See :doc:`data-preprocessor` for detail.
+
+.. program:: yp-preprocessor
+
+.. option:: file-names
```
Comment (DavidSimonin) [inline docs/cli.rst:85]: Shouls this be:
` .. option:: --input_file=FILENAME, -i FILENAME `

**Review by DavidSimonin — COMMENTED** [review]
I made few comments, but I wuld be happy eitherway for most of them.

File: docs/cli.rst (hunk 5 above)
Comment (mikecooke77) [reply docs/cli.rst:85]: Not changed see above comments

File: docs/cli.rst (hunk 4 above)
Comment (mikecooke77) [reply docs/cli.rst]: changed

File: src/yamlprocessor/datapreprocessor.py (hunk 1 above)
Comment (mikecooke77) [reply src/yamlprocessor/datapreprocessor.py:78]: This is always a required input and the current method follows the same method as yp-data.

File: src/yamlprocessor/datapreprocessor.py (hunk 2 above)
Comment (mikecooke77) [reply src/yamlprocessor/datapreprocessor.py]: agree added.

File: docs/data-preprocessor.rst (hunk 3 above)
Comment (mikecooke77) [reply docs/data-preprocessor.rst]: I take your point on the order and have updated this as suggested (apart from the -i as per comment above)

**Comment by DavidSimonin** [comment]
Happy with your comment above.

Regarding the the enviroment variable, should we use something like that to evalusate them instead of a mapping?:
```
>>> os.path.expandvars('$DATADIR/test/testing')
'/data/users/frsd/test/testing'
```

**Comment by mikecooke77** [comment]
> os.path.expandvars('$DATADIR/test/testing')

I like this as an idea but we are not passing them in as environment variables but as a defined list so I can't see how to get this to work.  Unless we want to define it as an environment variable which I would prefer not to do because the define method makes the user do this explicitly.

File: src/yamlprocessor/datapreprocessor.py (hunk 2 above)
Comment (matthewrmshin) [reply src/yamlprocessor/datapreprocessor.py]: Please use `-D` instead of `-d` for consistency with yp-data (and commands such as `cpp`).

**Comment by matthewrmshin** [comment]
See also yp-data's help on how environment variables are mapped.
* Environment variables are the default mapping for variable syntax in the file, unless the `--no-environment` option is specified.
* Each `-D KEY=VALUE` overrides.
* There is also a `--no-process-variable` option.

File: src/yamlprocessor/datapreprocessor.py (hunk 2 above)
Comment (mikecooke77) [reply src/yamlprocessor/datapreprocessor.py]: Changed to -D

**Comment by mikecooke77** [comment]
Thanks for the comments I have looked at the code on the `dataprocessor` and used this to update the `datapreprocessor` so that environment variables can also be used (and turned off). 

I don't think for the preprocessor we would ever need to no process variables.

**Review by DavidSimonin — APPROVED** [review]

**Comment by mikecooke77** [comment]
@matthewrmshin if you are happy with this code could we get it in the stack I can them start testing the jjdoc / jjaux split which will enable us to get up to date

**Review by matthewrmshin — APPROVED** [review]

**Pull request merged**